pkg/datamover/repository_server.go

```diff
@@ -0,0 +1,139 @@
+
+type repositoryServer struct {
+	outputName       string
+	repositoryServer *param.RepositoryServer
+	snapJSON         string
+	hostName         string
+}
+
+func (rs *repositoryServer) Pull(ctx context.Context, sourcePath, destinationPath string) error {
+	kopiaSnap, err := rs.unmarshalKopiaSnapshot()
+	if err != nil {
+		return err
+	}
+	password, err := rs.connectToKopiaRepositoryServer(ctx)
+	if err != nil {
+		return err
+	}
+	return kopiaLocationPull(ctx, kopiaSnap.ID, destinationPath, sourcePath, password)
+}
+
+func (rs *repositoryServer) Push(ctx context.Context, sourcePath, destinationPath string) error {
+	password, err := rs.connectToKopiaRepositoryServer(ctx)
+	if err != nil {
+		return err
+	}
+	return kopiaLocationPush(ctx, destinationPath, rs.outputName, sourcePath, password)
+}
+
+func (rs *repositoryServer) Delete(ctx context.Context, destinationPath string) error {
+	kopiaSnap, err := rs.unmarshalKopiaSnapshot()
+	if err != nil {
+		return err
+	}
+	password, err := rs.connectToKopiaRepositoryServer(ctx)
+	if err != nil {
+		return err
+	}
+	return kopiaLocationDelete(ctx, kopiaSnap.ID, destinationPath, password)
+}
+
+func (rs *repositoryServer) connectToKopiaRepositoryServer(ctx context.Context) (string, error) {
+	hostname, userPassphrase, err := rs.hostnameAndUserPassphrase()
+	if err != nil {
+		return "", errors.Wrap(err, "Error Retrieving Hostname and User Passphrase from Repository Server")
+	}
+
+	return userPassphrase, repository.ConnectToAPIServer(
+		ctx,
+		string(rs.repositoryServer.Credentials.ServerTLS.Data[kopia.TLSCertificateKey]),
+		userPassphrase,
+		hostname,
+		rs.repositoryServer.Address,
+		rs.repositoryServer.Username,
+		rs.repositoryServer.ContentCacheMB,
+		rs.repositoryServer.MetadataCacheMB,
+	)
+}
+
+func (rs *repositoryServer) unmarshalKopiaSnapshot() (*snapshot.SnapshotInfo, error) {
+	if rs.snapJSON == "" {
+		return nil, errors.New("kopia snapshot information is required to manage data using kopia")
+	}
+	kopiaSnap, err := snapshot.UnmarshalKopiaSnapshot(rs.snapJSON)
+	if err != nil {
+		return nil, err
+	}
+	return &kopiaSnap, nil
+}
+
+func (rs *repositoryServer) hostnameAndUserPassphrase() (string, string, error) {
+	var hostname, userPassphrase string
+	userCredsJSON, err := json.Marshal(rs.repositoryServer.Credentials.ServerUserAccess.Data)
+	if err != nil {
+		return "", "", errors.Wrap(err, "Error Unmarshalling User Credentials")
+	}
+	var userAccessMap map[string]string
+	if err := json.Unmarshal([]byte(string(userCredsJSON)), &userAccessMap); err != nil {
+		return "", "", errors.Wrap(err, "Failed to unmarshal User Credentials Data")
+	}
+
+	// check if hostname is provided in the repository server
+	if rs.hostName != "" {
+		hostname = rs.hostName
+		userPassphrase = userAccessMap[hostname]
+	} else {
+		// if hostname is not provided, use the first hostname in the map
+		for key, val := range userAccessMap {
+			hostname = key
+			userPassphrase = val
+			break
+		}
+	}
```

Comment:
```suggestion
        // use a random entry in the map as a default hostname 
	for key, val := range userAccessMap {
		hostname = key
		userPassphrase = val
		break
	}
	if rs.hostName != "" {
		hostname = rs.hostName
		userPassphrase = userAccessMap[hostname]
	}
```
else condition can be avoided

Reply:
Got it !